File: Tutorial4-8LayerImplementation2/src/main/java/com/test/tutorial4_8layerimplementation2/di/AppComponent.java
package com.test.tutorial4_8layerimplementation2.di;


import android.app.Application;

import com.test.tutorial4_8layerimplementation2.MainActivity;
import com.test.tutorial4_8layerlibrary.BaseActivity;
import com.test.tutorial4_8layerlibrary.di.BaseAppModule;
import com.test.tutorial4_8layerlibrary.di.BaseSubComponent;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Component(modules = {AppModule.class})
@Singleton
public interface AppComponent {

    BaseSubComponent.Builder baseComponentBuilder();

    @Component.Builder
    interface Builder {

        @BindsInstance
        Builder application(Application application);

        AppComponent build();

    }
}

File: Tutorial7-3BindsContextWithApplication/src/main/java/com/smarttoolfactory/tutorial7_3bindscontextwithapplication/di/AppComponent.kt
package com.smarttoolfactory.tutorial7_3bindscontextwithapplication.di

import android.app.Application
import com.smarttoolfactory.tutorial7_3bindscontextwithapplication.MainActivity
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [AppModule::class])
interface AppComponent {

    fun inject(mainActivity: MainActivity)

    @Component.Factory
    interface Factory {
        fun create(@BindsInstance application: Application): AppComponent
    }
}